pkg/auth/providers/activedirectory/activedirectory_client.go

```diff
@@ -136,14 +170,30 @@ func (p *adProvider) getPrincipalsFromSearchResult(result *ldapv2.SearchResult,
 	if config.GroupSearchBase != "" {
 		searchBase = config.GroupSearchBase
 	}
+
+	// Call common method for getting group principals
+	groupPrincipals, err = p.getGroupPrincipalsFromSearch(searchBase, nestedGroupsQuery, config, caPool, memberOf)
+	if err != nil {
+		return userPrincipal, groupPrincipals, err
+	}
+
+	return userPrincipal, groupPrincipals, nil
+
+}
+
+func (p *adProvider) getGroupPrincipalsFromSearch(searchBase string, filter string, config *v3.ActiveDirectoryConfig, caPool *x509.CertPool,
+	groupDN []string) ([]v3.Principal, error) {
+	var groupPrincipals []v3.Principal
+	var nilPrincipal []v3.Principal
+
 	search := ldapv2.NewSearchRequest(searchBase,
 		ldapv2.ScopeWholeSubtree, ldapv2.NeverDerefAliases, 0, 0, false,
-		nestedGroupsQuery,
+		filter,
 		ldap.GetGroupSearchAttributes(MemberOfAttribute, ObjectClassAttribute, config), nil)
 
 	lConn, err := p.ldapConnection(config, caPool)
```

Comment:
We need to close this ldap connection after the method exits

Reply:
done